### PR TITLE
api, server, ui: allow host auto-select for migrateVirtualMachineWith…

### LINKED ISSUE #6773: AutoSelect live migration fails with local storage
##### ISSUE TYPE
 * Bug Report

##### COMPONENT NAME
~~~
UI, API
~~~

##### CLOUDSTACK VERSION
~~~
4.16.1
~~~

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
N/A

##### SUMMARY
When live migrating an instance and choosing "AutoSelect" it is not possible to live migrate an instance with local storage, while when selecting the destination host, it is possible to live migrate an instance with local storage.

##### STEPS TO REPRODUCE
~~~
- Spawn vm on local storage
- Live migrate instance in UI using "AutoSelect"
- Migration fails

CS calls migrateVirtualMachine instead of migrateVirtualMachineWithVolume
~~~

##### EXPECTED RESULTS
~~~
CloudStack to be smart enough to also live migrate instances with local storage when using "AutoSelect"
CS calls migrateVirtualMachineWithVolume instead of migrateVirtualMachine
~~~

##### ACTUAL RESULTS

- Error is returned:
```
Migration failed
Unsupported operation, VM uses Local storage, cannot migrate
```

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #6773

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@DaanHoogland please note I've not done any sort of testing for this yet

**Comment by DaanHoogland** [comment]
> @DaanHoogland please note I've not done any sort of testing for this yet

I'll do some when I get around. Do you have any test advice on this? I couldn´t reproduce the  issue on 4.17.

**Comment by shwstppr** [comment]
@DaanHoogland To reproduce, I think one needs to just select Automatically assign in the migrate instance form for a vm that has volumes on local storage,
Screenshot from 2022-09-30 15-52-43
For tests,
- we will need to test this case which I feel would fail :-D
- test for any regression - user vm migration, systemvm migration, vm storage migration using migrateVirtualMachineWithVolume API and maybe cross-cluster migration

**Review by weizhouapache — COMMENTED** [review]
Nice enhancement

**Comment by DaanHoogland** [comment]
> Nice enhancement

Did you test @weizhouapache ? I didn´t get to it yet.

**Comment by weizhouapache** [comment]
> > Nice enhancement
> 
> Did you test @weizhouapache ? I didn´t get to it yet.

I have not tested it yet

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java [hunk 1]
```diff
@@ -83,6 +83,12 @@ public class MigrateVirtualMachineWithVolumeCmd extends BaseAsyncCmd {
                "<1b331390-59f2-4796-9993-bf11c6e76225>&migrateto[2].pool=<41fdb564-9d3b-447d-88ed-7628f7640cbc>")
     private Map migrateVolumeTo;
 
+    @Parameter(name = ApiConstants.AUTO_SELECT,
+            since = "4.18.0",
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java:87]: no longer true

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -6751,8 +6737,23 @@ private boolean isImplicitPlannerUsedByOffering(long offeringId) {
         return implicitPlannerUsed;
     }
 
-    private boolean isVmVolumesOnZoneWideStore(VMInstanceVO vm) {
-        final List<VolumeVO> volumes = _volsDao.findCreatedByInstance(vm.getId());
+    private boolean isVmVolumesUsingLocalStorage(final List<VolumeVO> volumes) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6740]: The code in this method looks exactly like the bit in `isVmVolumesOnZoneWideStore`

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 3]
```diff
@@ -6751,8 +6737,23 @@ private boolean isImplicitPlannerUsedByOffering(long offeringId) {
...
+            return false;
+        }
+        for (Volume volume : volumes) {
+            if (volume == null || volume.getPoolId() == null) {
+                return false;
+            }
+            StoragePoolVO pool = _storagePoolDao.findById(volume.getPoolId());
+            if (pool == null || !ScopeType.ZONE.equals(pool.getScope())) {
+                return false;
+            }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6751]: should this read something like 
```suggestion
            if (volume == null || volume.getPoolId() == null) {
                return false;
            }
            StoragePoolVO pool = _storagePoolDao.findById(volume.getPoolId());
            if (pool == null || !ScopeType.ZONE.equals(pool.getScope())) {
                return false;
            }
```
???

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 4]
```diff
@@ -6952,6 +6958,18 @@ public VirtualMachine migrateVirtualMachineWithVolume(Long vmId, Host destinatio
...
+            Long poolId = null;
+            if (MapUtils.isNotEmpty(volToPoolObjectMap)) {
+                poolId = new ArrayList<>(volToPoolObjectMap.values()).get(0);
+            }
+            DeployDestination deployDestination = chooseVmMigrationDestination(vm, srcHost, poolId);
+            if (deployDestination == null) {
+                throw new CloudRuntimeException("Unable to find suitable destination to migrate VM " + vm.getInstanceName());
+            }
+            destinationHost = deployDestination.getHost();
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6971]: can this be extracted?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
Recreated here https://github.com/apache/cloudstack/pull/7554

**Pull request closed without merging**